File: stripe/src/main/java/com/stripe/android/net/StripeResponse.java
package com.stripe.android.net;

import android.support.annotation.Nullable;

import java.util.List;
import java.util.Map;

/**
 * Represents a response from the Stripe servers.
 */
public class StripeResponse {

    private String mResponseBody;
    private int mResponseCode;
    private Map<String, List<String>> mResponseHeaders;

    /**
     * Object constructor.
     *
     * @param responseCode the response code (i.e. 404)
     * @param responseBody the body of the response
     * @param responseHeaders any headers associated with the response
     */
    public StripeResponse(
            int responseCode,
            String responseBody,
            @Nullable Map<String, List<String>> responseHeaders) {
        mResponseCode = responseCode;
        mResponseBody = responseBody;
        mResponseHeaders = responseHeaders;
    }

    /**
     * @return the {@link #mResponseCode response code}.
     */
    public int getResponseCode() {
        return mResponseCode;
    }

    /**
     * @return the {@link #mResponseBody response body}.
     */
    public String getResponseBody() {
        return mResponseBody;
    }

    /**
     * @return the {@link #mResponseHeaders response headers}.
     */
    @Nullable
    public Map<String, List<String>> getResponseHeaders() {
        return mResponseHeaders;
    }
}

File: payments-core/src/main/java/com/stripe/android/core/networking/StripeResponse.kt
package com.stripe.android.core.networking

import androidx.annotation.RestrictTo
import java.io.File
import java.net.HttpURLConnection
import java.net.HttpURLConnection.HTTP_MULT_CHOICE

/**
 * Represents a response from the Stripe servers.
 * Upon receiving the HTTP response, its body is parsed into [ResponseBody], such as a
 * [String] or a [File].
 *
 * @param code the response code (i.e. 404)
 * @param body the body of the response
 * @param headers any headers associated with the response
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
data class StripeResponse<ResponseBody> internal constructor(
    /**
     * the response code
     */
    internal val code: Int,
    /**
     * the response body
     */
    internal val body: ResponseBody?,
    /**
     * the response headers
     */
    internal val headers: Map<String, List<String>> = emptyMap()
) {
    internal val isOk: Boolean = code == HttpURLConnection.HTTP_OK
    internal val isError: Boolean = code < HttpURLConnection.HTTP_OK || code >= HTTP_MULT_CHOICE
    internal val isRateLimited = code == HTTP_TOO_MANY_REQUESTS

    internal val requestId: RequestId? = RequestId.fromString(
        getHeaderValue(HEADER_REQUEST_ID)?.firstOrNull()
    )

    override fun toString(): String {
        return "$HEADER_REQUEST_ID: $requestId, Status Code: $code"
    }

    internal fun getHeaderValue(key: String): List<String>? {
        return headers.entries
            .firstOrNull {
                it.key.equals(key, ignoreCase = true)
            }?.value
    }

    internal companion object {
        internal const val HEADER_REQUEST_ID = "Request-Id"
    }
}
